### ISSUE TITLE
Validate is called on Model.get

### ISSUE THREAD

**Issue by Tirke** [dynamoosejs/dynamoose#583]
### Summary:

`Model.get()` is calling validate on attributes but I don't think it's necessary (and it's breaking our way to validate complex models). I've written a test showing this behavior.

### Code sample:
```javascript
const model = dynamoose.model('no-validate', new dynamoose.Schema(
      {
        id: {
          type: String,
          required: true,
          hashKey: true,
          validate: function (val, model) {
            if (!model.name){
              // Is called on get :(
              throw new Error('should not throw');
            }
            return true;
          }
        },
        projectId: {
          type: String,
          required: true,
          rangeKey: true,
          index: {
            global: true,
            name: 'projectId-index',
            project: true,
          },
        },
      }));


    let error, res;
    try {
      await model.create({id: 'idid', projectId:'project', name: 'Rob'});
      res = await model.get({id: 'idid', projectId: 'project'});
    } catch (e) {
      error = e;
    }
    should.not.exist(error);
    res.id.should.eql(1);
    res.name.should.eql('Rob');
```

### Current output and behavior:

Validate is called when calling `Model.get()`

### Expected output and behavior:

Validate shouldn't be called when calling `Model.get()`

### Environment:

Operating System: MacOS Mojave
Operating System Version: latest
Node.js version (`node -v`): 8.15.0
NPM version: (`npm -v`): 6.4.1
Dynamoose version: latest
Dynamoose Plugins: none

-

### Dynamoose Plugins:

- [ ] Yes, I believe that one or multiple 3rd party Dynamoose plugins are effecting this issue
- [ ] No, I believe this issue is independent of any 3rd party Dynamoose plugins I'm using
- [ ] Unknown, I'm unsure if Dynamoose plugins are effecting this issue
- [x] I am not using any Dynamoose plugins

### Type (select 1):
- [x] Bug report
- [ ] Feature suggestion
- [ ] Question
- [ ] Other suggestion
- [ ] Something not listed here

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x] I have tested the code provided and am confident it doesn't work as intended
- [x] I have ensured that all of my plugins that I'm using with Dynamoose are listed above
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
@Tirke Why do you think that `validate` shouldn't be called on `Model.get`?

**Comment by Tirke** [comment]
I'm trying to retrieve data in the database, this will never change state in the database so no validation should occur (IMO). Maybe there will be some errors because I forgot some fields in my query but dynamo will throw and tell me this already.
 
We have a use case for complex validation inside only one `validate()` function.
We use Joi to validate our entire model inside the id field `validate()` (using the second parameter) and so we only use dynamoose schema to declare the necessary index fields.
We need to do this because currently dynamoose is lacking complex validation capabilities like this :

```javascript
 config: Joi.alternatives()
    .when('type', {
      is: SFTP,
      then: Joi.object()
        .keys({
          pollingFrequency: Joi.number()
            .integer()
            .required(),
          host: Joi.string().required(),
          port: Joi.string().required(),
          password: Joi.string(),
          privateKey: Joi.string(),
          remoteDirectory: Joi.string().required(),
          user: Joi.string().required(),
        })
        .xor('password', 'privateKey'),
    })
    .when('type', {
      is: SIGFOX,
      then: Joi.object().keys({
        groupId: Joi.string(),
        login: Joi.string(),
        password: Joi.string(),
        contractId: Joi.string(),
      }),
    })
...
```
To summarize I'm trying to have a dynamic schema based on the value of some other fields in my schema. Eg when type is X then config will contain Y, Z keys. When type is Y then config will contain K, T keys.

I hope this is clear enough.

**Comment by fishcharlie** [comment]
@Tirke Yeah I personally disagree. The whole point of a modeling tool like Dynamoose is being strict and giving structure to your data. Some users might rely on validation occurring on everything. And I think that is a fair assumption to make. If data is invalid, it's invalid, regardless if it's a get or put. If the point is to be strict, why should we give you bad data?

It almost looks like you are wanting dynamic schemas for each item. The only thing I can suggest at this point is the `saveUnknown` property.

**Comment by fishcharlie** [comment]
@Tirke Can you confirm what the Mongoose behavior of this is?

**Comment by Tirke** [comment]
I don't know Mongoose well enough to answer that question with 100% certitude but a quick look into documentation gave me :

- Validation is defined in the SchemaType
- Validation is middleware. Mongoose registers validation as a pre('save') hook on every schema by default.
- You can manually run validation using doc.validate(callback) or doc.validateSync()
- Validators are not run on undefined values. The only exception is the required validator.
- Validation is asynchronously recursive; when you call Model#save, sub-document validation is executed as well. If an error occurs, your Model#save callback receives it
- Validation is customizable

So I'm expecting validation on save or update only.

I'm not sure I can agree on validating on a simple get. Keep in mind that the current behavior is done BEFORE the get request. There's basically nothing to validate. I just want to retrieve data why should I validate the request to retrieve data? But if it's what user really want so be it 🤷‍♂️ 

BTW if people are wondering, we are now doing validation in our services (so outside dynamoose modeling). This way we can control exactly when it is called. Another good way of doing this would be to use dynamoose hooks when it's stable.

**Comment by hweeks** [comment]
Let's dig into the findOne method in mongoose, as you can see it inspired a lot of what is happening here. In that method no validation occurs at all, it just queries based on the conditions it builds. check [this](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L1916-L1947).

If we then take a look at the create method, we see that is the only place validation occurs. [the method](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L2734-L2818) and [the validation](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L2788)

This shows that mongoose only validates on create, never on fetch. That said, I know we've had a bit of push back on going for full parity with mongoose, and that is not even claimed as a goal in the project.

Personally, I'm all for it as it will help with performance to reduce the callback overhead on a per call basis a lot.

@fishcharlie

**Comment by stewartjarod** [comment]
I agree. Validating pre-fetching doesn't really make sense. Dynamo will throw an appropriate error if there is an issue as well.

**Comment by fishcharlie** [comment]
This behavior is being changed in version 2.0 and should no longer call validate on retrieval requests. Although there are no automated tests to confirm this and this behavior is not documented in 2.0. There is a chance I’ll get around to that before 2.0 gets merged, but if it’s important to you I’d suggest making a PR to the 2.0 branch and add some automated tests surrounding this with maybe some documentation. Otherwise can’t guarantee that this won’t change in the future without a breaking release.

**Comment by dolsem** [comment]
@fishcharlie unless I'm missing the point, perhaps it's worth making validation optional? The big difference between relational databases and DynamoDB is that you can technically store documents that don't conform to a single schema in a single table. So I feel like if there's a monolithic app solely responsible for database writes it makes sense to disable validation, but if you use something like microservices for updates and expect potential irregularity of your data (worst case if a regression occurs), it's probably worth it to keep validation on.

**Comment by fishcharlie** [comment]
@dolsem What do you believe should be the default behavior for this?

**Comment by dolsem** [comment]
I can see arguments for both. Personally, I would prefer it to be on by default, but more importantly, it would be nice to stick it into various schema examples throughout the docs, to make sure users have seen the configuration option and are aware of it. Generally speaking, I like defaults to have minimal overhead, but extra safeguards are worth it.
